coderd/workspaceusage/tracker.go

```diff
@@ -0,0 +1,181 @@
+}
+
+// Tracker tracks and de-bounces updates to workspace usage activity.
+// It keeps an internal map of workspace IDs that have been used and
+// periodically flushes this to its configured Store.
+type Tracker struct {
+	log      slog.Logger            // you know, for logs
+	mut      sync.Mutex             // protects m
+	m        map[uuid.UUID]struct{} // stores workspace ids
+	s        Store                  // for flushing data
+	tickCh   <-chan time.Time       // controls flush interval
+	stopTick func()                 // stops flushing
+	stopCh   chan struct{}          // signals us to stop
+	stopOnce sync.Once              // because you only stop once
+	doneCh   chan struct{}          // signifies that we have stopped
+	flushCh  chan int               // used for testing.
+}
+
+// New returns a new Tracker. It is the caller's responsibility
+// to call Close().
+func New(s Store, opts ...Option) *Tracker {
+	hb := &Tracker{
+		log:      slog.Make(sloghuman.Sink(os.Stderr)),
+		m:        make(map[uuid.UUID]struct{}, 0),
+		s:        s,
+		tickCh:   nil,
+		stopTick: nil,
+		stopCh:   make(chan struct{}),
+		doneCh:   make(chan struct{}),
+		flushCh:  nil,
+	}
+	for _, opt := range opts {
+		opt(hb)
+	}
+	if hb.tickCh == nil && hb.stopTick == nil {
+		ticker := time.NewTicker(DefaultFlushInterval)
+		hb.tickCh = ticker.C
+		hb.stopTick = ticker.Stop
+	}
+	return hb
+}
+
+type Option func(*Tracker)
+
+// WithLogger sets the logger to be used by Tracker.
+func WithLogger(log slog.Logger) Option {
+	return func(h *Tracker) {
+		h.log = log
+	}
+}
+
+// WithFlushInterval allows configuring the flush interval of Tracker.
+func WithFlushInterval(d time.Duration) Option {
+	return func(h *Tracker) {
+		ticker := time.NewTicker(d)
+		h.tickCh = ticker.C
+		h.stopTick = ticker.Stop
+	}
+}
+
+// WithFlushChannel allows passing a channel that receives
+// the number of marked workspaces every time Tracker flushes.
+// For testing only.
+func WithFlushChannel(c chan int) Option {
+	return func(h *Tracker) {
+		h.flushCh = c
+	}
+}
+
+// WithTickChannel allows passing a channel to replace a ticker.
+// For testing only.
+func WithTickChannel(c chan time.Time) Option {
+	return func(h *Tracker) {
+		h.tickCh = c
+		h.stopTick = func() {}
+	}
+}
+
+// Add marks the workspace with the given ID as having been used recently.
+// Tracker will periodically flush this to its configured Store.
+func (wut *Tracker) Add(workspaceID uuid.UUID) {
+	wut.mut.Lock()
+	wut.m[workspaceID] = struct{}{}
+	wut.mut.Unlock()
+}
+
+// flushLocked updates last_used_at of all current workspace IDs.
+// MUST HOLD LOCK BEFORE CALLING
+func (wut *Tracker) flushLocked(now time.Time) {
+	if wut.mut.TryLock() {
+		panic("developer error: must lock before calling flush()")
+	}
+	count := len(wut.m)
+	defer func() { // only used for testing
+		if wut.flushCh != nil {
+			wut.flushCh <- count
+		}
+	}()
```

Comment:
Only run the go routine if flushCh exists
```suggestion
	if wut.flushCh != nil { // only used for testing
		defer func() {
			wut.flushCh <- count
		}()
	}
```

Reply:
Good catch!